File: src/main/java/svnserver/ext/gitea/mapping/GiteaMapper.java
/**
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.ext.gitea.mapping;

import java.io.IOException;
import java.util.Iterator;
import java.util.LinkedList;

import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tmatesoft.svn.core.SVNException;

import io.gitea.ApiClient;
import io.gitea.ApiException;
import io.gitea.api.RepositoryApi;
import io.gitea.model.Repository;


public class GiteaMapper extends Thread implements DirectoryWatcher.DirectoryMapping {
  @NotNull
  private static final Logger log = LoggerFactory.getLogger(GiteaMapper.class);

  private LinkedList<String> toAdd = new LinkedList<>();
  private LinkedList<String> toRemove = new LinkedList<>();
  private GiteaMapping mapping;
  private RepositoryApi repositoryApi;

  public GiteaMapper(ApiClient apiClient, GiteaMapping mapping) {
    this.repositoryApi = new RepositoryApi(apiClient);
    this.mapping = mapping;
    this.start();
  }

  public void run() {
    try {
      while (true) {
        synchronized(toAdd) {
          for (Iterator<String> it = toRemove.iterator(); it.hasNext(); ) {
            String projectName = it.next();
            mapping.removeRepository(projectName);
            it.remove();
          }
          for (Iterator<String> it = toAdd.iterator(); it.hasNext();) {
            String projectName = it.next();
            String owner = projectName.substring(0, projectName.indexOf('/'));
            String repo = projectName.substring(projectName.indexOf('/') + 1);
            try {
              Repository repository = repositoryApi.repoGet(owner, repo);
              GiteaProject project = mapping.addRepository(repository);
              if (project != null) {
                project.initRevisions();
                it.remove();
              }
            } catch (ApiException e) {
              // Not ready yet - try again later...
            } catch (IOException | SVNException e) {
              log.error("Processing error whilst adding repository: {} / {}: {}", owner, repo, e.getMessage());
            }
          }
        }
        sleep(1000);
      }
    } catch(InterruptedException e) {
      return;
    }
  }

  @Override
  public void addRepository(String owner, String repo) {
    synchronized(toAdd) {
      toAdd.add(owner + "/" + repo);
    }
  }

  @Override
  public void removeRepositories(String owner) {
    synchronized(toAdd) {
      for (GiteaProject project : mapping.getRepositories()) {
        if (project.getOwner().equals(owner)) {
          toRemove.add(project.getRepositoryName());
          toAdd.remove(project.getRepositoryName());
        }
      }
    }
  }

  @Override
  public void removeRepository(String owner, String repo) {
    synchronized(toAdd) {
      toRemove.add(owner + "/" + repo);
      toAdd.remove(owner + "/" + repo);
    }
  }
}

File: src/main/kotlin/svnserver/ext/gitea/mapping/GiteaMapper.kt
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver.ext.gitea.mapping

import io.gitea.ApiClient
import io.gitea.ApiException
import io.gitea.api.RepositoryApi
import org.tmatesoft.svn.core.SVNException
import svnserver.Loggers
import svnserver.ext.gitea.mapping.DirectoryWatcher.DirectoryMapping
import java.io.IOException
import java.util.*
import java.util.concurrent.locks.ReentrantLock

internal class GiteaMapper(apiClient: ApiClient, private val mapping: GiteaMapping) : Thread(), DirectoryMapping {
    private val lock = ReentrantLock()
    private val toAdd = LinkedList<String>()
    private val toRemove = LinkedList<String>()
    private val repositoryApi: RepositoryApi = RepositoryApi(apiClient)
    override fun run() {
        try {
            while (true) {
                lock.lock()
                try {
                    run {
                        val it = toRemove.iterator()
                        while (it.hasNext()) {
                            val projectName = it.next()
                            mapping.removeRepository(projectName)
                            it.remove()
                        }
                    }
                    val it = toAdd.iterator()
                    while (it.hasNext()) {
                        val projectName = it.next()
                        val owner = projectName.substring(0, projectName.indexOf('/'))
                        val repo = projectName.substring(projectName.indexOf('/') + 1)
                        try {
                            val repository = repositoryApi.repoGet(owner, repo)
                            val project = mapping.addRepository(repository)
                            project?.initRevisions()
                            it.remove()
                        } catch (e: ApiException) {
                            // Not ready yet - try again later...
                        } catch (e: IOException) {
                            log.error("Processing error whilst adding repository: {} / {}: {}", owner, repo, e.message)
                        } catch (e: SVNException) {
                            log.error("Processing error whilst adding repository: {} / {}: {}", owner, repo, e.message)
                        }
                    }
                } finally {
                    lock.unlock()
                }
                sleep(1000)
            }
        } catch (e: InterruptedException) {
            // noop
        }
    }

    override fun addRepository(owner: String, repo: String) {
        lock.lock()
        try {
            toAdd.add("$owner/$repo")
        } finally {
            lock.unlock()
        }
    }

    override fun removeRepositories(owner: String) {
        lock.lock()
        try {
            for (project in mapping.mapping.values) {
                if (project.owner == owner) {
                    toRemove.add(project.repositoryName)
                    toAdd.remove(project.repositoryName)
                }
            }
        } finally {
            lock.unlock()
        }
    }

    override fun removeRepository(owner: String, repo: String) {
        lock.lock()
        try {
            toRemove.add("$owner/$repo")
            toAdd.remove("$owner/$repo")
        } finally {
            lock.unlock()
        }
    }

    companion object {
        private val log = Loggers.gitea
    }

    init {
        start()
    }
}
